### PR TITLE
added check for NumberType

### LINKED ISSUE #44107: TorchScript number type is cast to an integer during assignment inside jit.trace
## 🐛 Bug

The TorchScript `number` type gets cast to an integer type during value assignment inside jit.trace

## To Reproduce

This is the minimal code example that produces this bug.

```python
import torch


@torch.jit.script
def test(some_tensor: torch.Tensor):
    for loop_idx in range(some_tensor.shape[0]):
        temp = some_tensor[loop_idx].item()
        print(temp)
        some_tensor[loop_idx] = temp

    return some_tensor


some_tensor = torch.Tensor([0.5, 1.5])
print(some_tensor)
print(test(some_tensor), '\n')
print(test.code)
```

Output:

```
tensor([0.5000, 1.5000])
0.5
1.5
tensor([0., 1.])

def test(some_tensor: Tensor) -> Tensor:
  for loop_idx in range((torch.size(some_tensor))[0]):
    _0 = torch.select(some_tensor, 0, loop_idx)
    temp = torch.item(_0)
    print(temp)
    _1 = torch.select(some_tensor, 0, loop_idx)
    _2 = torch.copy_(_1, int(temp))
  return some_tensor
```

This happens with `float16` and `float64` data types, as well as with `cpu` and `cuda` tensors .

## Expected behavior

The internal number type being cast to the proper type of the tensor it's being assigned to.

This doesn't happen when `.item()` is left out, since `temp` remains a `Tensor` and preserves its type, or when explicit type casting is used, i.e. `some_tensor[loop_idx] = float(temp)` and `some_tensor[loop_idx] = torch.jit.annotate(float, temp)`

This might also be related to other bugs that produce different outputs, that have `.item()` inside a `torch.jit.script` function, like in the case of summing multiple loss scalar values.

## Environment

```sh
PyTorch version: 1.6.0
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: Ubuntu 18.04.4 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.10.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 2080
Nvidia driver version: 450.51.05
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] torch==1.6.0
[pip3] torchvision==0.7.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.2.89              hfd86e86_1  
[conda] mkl                       2020.1                      217  
[conda] mkl-service               2.3.0            py37he904b0f_0  
[conda] mkl_fft                   1.1.0            py37h23d657b_0  
[conda] mkl_random                1.1.1            py37h0da4684_0    conda-forge
[conda] numpy                     1.19.1           py37hbc911f0_0  
[conda] numpy-base                1.19.1           py37hfa32c7d_0  
[conda] pytorch                   1.6.0           py3.7_cuda10.2.89_cudnn7.6.5_0    pytorch
[conda] torchvision               0.7.0                py37_cu102    pytorch
```

cc @gmagogsfm

### PR DESCRIPTION
**Pull request by hlahkar**
Fixes #44107

### CODE REVIEW

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 1]
```diff
@@ -2175,9 +2175,12 @@ struct to_ir {
     // Used in emitSubscriptAssign to convert:
     //   `tensor(...)[x] = 99` to `tensor(...)[x] = tensor(99)`
     // Mirrors the `valueToTensor` behavior in python_variable_indexing.cpp
+    //adding implicit conversion for NumberType also; else it float values
```
Comment (VinodSKumar) [inline torch/csrc/jit/frontend/ir_emitter.cpp:2178]: Missing space after //

**Review by VinodSKumar — COMMENTED** [review]
Also can add a test case for this scenario

**Review by bhosmer — COMMENTED** [review]
LGTM! @eellison pinging you (or PYMK :) for a quick check for subtleties - I haven't worked in the emitter specifically, but given the other typechecks, it sure looks like the right place to me.

Also though, any suggestions on where @hlahkar might add the repro from #44107 as a test? If there are similar tests for the other typechecks, it would be a nice addition.

**Review by eellison — APPROVED** [review]
LGTM. A test would be nice. Thanks for the PR!
